plugins/user.py

```diff
@@ -0,0 +1,507 @@
+            return
+
+        user.revoke_permissions(requested_permissions)
+        user.save()
+
+        self.direct_reply(message, "New permissions for '{}' are: {} ".format(nick, ', '.join(user.permissions)))
+
+
+class User(object):
+    """
+    Represents a user.
+
+    Contains sufficient information to determine if they are authenticated and what permissions they have.
+
+    Constructor Arguments:
+
+        plugin: The plugin that is currently responding to a message.
+        user_metadata: A dictionary that contains a "nick" key and a "hipchat_id". This dictionary is used by will to
+            represent a user quite frequently.
+        token: The secret base32 TOTP token used to verify transient tokens provided by external devices. This is as
+            sensitive as a user's password and should be handled with care.
+        permissions: A set() containing strings that represent all permissions the user has been granted.
+        verified_time: The last time the user verified their identity as a datetime object with tzinfo == UTC.
+    """
+
+    # Design notes
+
+    # All of the data about that user that needs to be persisted is stored in a dictionary in the key-value store. Each
+    # user has their own key. Since the bot may be responding to several requests at once on different threads, this
+    # seemed safer than having one massive dictionary storing all user information in it (fewer race conditions). That
+    # said - some race conditions to exists, particularly around permission management. If two admins decide to modify
+    # permissions of a particular user at roughly the same time, then one of them will overwrite the other which may
+    # not be the desired effect. This could be mitigated by taking advantage of some of Redis's built in "set"
+    # management features.
+
+    # From a performance standpoint, some calls here require calls to the hipchat API. I know (for example) that
+    # retrieving the user's timezone requires one such call. I suspect others do as well, however, the calls are made
+    # deep inside Will, so it's not abundantly clear which function calls result in API calls and which ones don't.
+    # If there were performance problems, or we start hitting rate limits on the hipchat API we could cache stuff like
+    # the timezone in Redis, and just assume it's very "slow changing".
+
+    DEFAULT_SESSION_DURATION = 30 * 60  # 30 minutes in seconds
+    QR_CODE_VALIDITY_DURATION = 10 * 60  # 10 minutes in seconds
+
+    def __init__(self, plugin, user_metadata, token, permissions, verified_time=None):
+        self.nick = user_metadata['nick']
+        self.hipchat_id = user_metadata['hipchat_id']
+        self.totp = pyotp.TOTP(token)
+        self.token = token
+        self.permissions = permissions
+        self.plugin = plugin
+        self.verified_time = verified_time
+        self.session_duration = getattr(settings, 'TWOFACTOR_SESSION_DURATION', self.DEFAULT_SESSION_DURATION)
```

Comment:
It isn't... I'm just caching the value so that I don't have to copy paste that awful code snippet around.